### PR TITLE
Checkable: Don't recalculate `next_check` for remotely generated `cr`

### PR DESCRIPTION
**Pull request by yhabteab**
Currently, when processing a `CheckResult`, it will first trigger an `OnNextCheckChanged` event, which is sent to all connected endpoints. Then, when `Checkable#ProcessCheckResult()` returns, an `OnNewCheckResult` event is fired, which is of course also sent to all connected endpoints.

Next, the other endpoints receive the `event::SetNextCheck` cluster event followed by `event::CheckResult`and invoke
`checkable#SetNextCheck()` and `Checkable#ProcessCheckResult()` with the newly received check. So they also try to recalculate the next check themselves and invalidate the previously received next check timestamp from the source endpoint. Since each endpoint calculates it relative to **time#now** (recomputing the next check relative to the `last check/cr#schedule_end` does not work for active checks either, as each endpoint randomly initialises its own scheduling offset), the recalculated next check will always differ by a split second/millisecond on each of them. As a consequence, two Icinga DB HA instances will generate two different checksums for the same state and causes the state ~histories~ tables to be fully resynchronised after a takeover/Icinga 2 reload.

### Before

```diff
~ diff <(curl -sSku root:icinga 'https://localhost:5666/v1/objects/hosts/icinga2?pretty=1') <(curl -sSku root:icinga 'https://localhost:5665/v1/objects/hosts/icinga2?pretty=1')
94,95c94,95
<                 "next_check": 1709202335.3899999,
<                 "next_update": 1709202350.4348836,
---
>                 "next_check": 1709202334.593121,
>                 "next_update": 1709202349.6380048,
99,100c99,100
<                 "package": "_cluster",
<                 "paused": false,
---
>                 "package": "_etc",
>                 "paused": true,
110c110
<                     "path": "/Users/yhabteab/Workspace/icinga2-2/prefix/var/lib/icinga2/api/zones/master/_etc/hosts.conf"
---
>                     "path": "/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/zones.d/master/hosts.conf"
```

### After

```diff
~ diff <(curl -sSku root:icinga 'https://localhost:5666/v1/objects/hosts/icinga2?pretty=1') <(curl -sSku root:icinga 'https://localhost:5665/v1/objects/hosts/icinga2?pretty=1')
99,100c99,100
<                 "package": "_cluster",
<                 "paused": false,
---
>                 "package": "_etc",
>                 "paused": true,
110c110
<                     "path": "/Users/yhabteab/Workspace/icinga2-2/prefix/var/lib/icinga2/api/zones/master/_etc/hosts.conf"
---
>                     "path": "/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/zones.d/master/hosts.conf"
```

### CODE REVIEW

File: lib/icinga/checkable-check.cpp [hunk 1]
```diff
@@ -360,9 +360,14 @@ Checkable::ProcessingResult Checkable::ProcessCheckResult(const CheckResult::Ptr
...
-		UpdateNextCheck(origin);
-	} else {
+	// Don't recompute the next check when the current check isn't generated by this endpoint. When the check is
+	// remotely generated we should've already received the "SetNextCheck" event before the "event::CheckResult"
+	// cluster event. Otherwise, the next check received before this check will be invalidated and cause the Checkable
+	// "next_check/next_update" in a HA setup to always be different from the other endpoint as the "m_SchedulingOffset"
+	// is randomly initialised on each node.
+	if (!origin && cr->GetActive()) {
+		UpdateNextCheck();
+	} else if (!origin) {
```
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/icinga/checkable-check.cpp:370]: In contrast, the way you did it is a bit confusing.

**Review by Al2Klimov — COMMENTED** [review]
Does this work with CRs from command endpoints?

**Comment by Al2Klimov** [comment]
A command endpoint is NOT in the zone of the checkable it checks. So it would send a 'next check changed' message along with the CR, but the master would ignore it: _Discarding 'next check changed' message for checkable '...' from '...': Unauthorized access_ And the master wouldn't update the next check by itself as _!origin_ is false. The CR's origin is the command endpoint.

**Comment by julianbrost** [comment]
> A command endpoint is NOT in the zone of the checkable it checks.

It may or may not be in that zone. You can also use `command_endpoint` to pin the check execution to a particular node within a HA zone.

**Comment by Al2Klimov** [comment]
* ClusterEvents::CheckResultAPIHandler() allows checkable zone, its parents and the command endpoint: https://github.com/Icinga/icinga2/blob/9e31b8b5590c6d67b7dd538b2c884bd377a4e486/lib/icinga/clusterevents.cpp#L171
* ClusterEvents::NextCheckChangedAPIHandler() allows checkable zone and its parents: https://github.com/Icinga/icinga2/blob/9e31b8b5590c6d67b7dd538b2c884bd377a4e486/lib/icinga/clusterevents.cpp#L236

**Comment by yhabteab** [comment]
> * ClusterEvents::CheckResultAPIHandler() allows checkable zone, its parents and the command endpoint:

But that receiver doesn't pass the `origin` to `ProcessCheckResult()` if the message is a result of a command endpoint.
https://github.com/Icinga/icinga2/blob/9e31b8b5590c6d67b7dd538b2c884bd377a4e486/lib/icinga/clusterevents.cpp#L178-L179

**Comment by yhabteab** [comment]
Master <-> satellite:
```diff
~/Workspace/icinga2 (next-check-cluster-sync-issue ✗) diff <(curl -sSku root:icinga 'https://localhost:5667/v1/objects/services/satellite!cluster?pretty=1') <(curl -sSku root:icinga 'https://localhost:5666/v1/objects/services/satellite!cluster?pretty=1')
223c223
<                 "package": "_etc",
---
>                 "package": "_cluster",
234c234
<                  "path": "/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/zones.d/satellite/host.conf"
---
>                  "path": "/Users/yhabteab/ClionProjects/icinga2/prefix/var/lib/icinga2/api/zones/satellite/_etc/host.conf"
```

**Review by Al2Klimov — APPROVED** [review]
Command Endpoints aren't problematic!

https://github.com/Icinga/icinga2/blob/bca1a8447af26313123183efa863f947f9590731/lib/icinga/clusterevents.cpp#L178-L181

**Comment by julianbrost** [comment]
> As a consequence, two Icinga DB HA instances will generate two different checksums for the same state and causes the **state histories to be fully resynchronised** after a takeover/Icinga 2 reload.

This should not be related to the host/service history and there's no full resynchronization of that either. I think you mean the host/service state tables.

**Review by julianbrost — APPROVED** [review]

**Pull request merged**